### PR TITLE
Prevent IrisInConfigurationSpaceFromCliqueCover from crashing the whole program if IrisInConfigurationSpace fails

### PR DESCRIPTION
**Pull request by AlexandreAmice**
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376)

### CODE REVIEW

**Review by AlexandreAmice — COMMENTED** [review]
This is a minimal fix for [#21343](https://github.com/RobotLocomotion/drake/issues/21343). We have a more graceful fix with @wernerpe which we will implement in the future, but it is part of a larger internal overhaul.

+@wernerpe for feature review.
cc @RussTedrake Can you try this on your example that was failing?

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376#-:-Nwp6O0o7R5qOUlcEsR_:btgr2zs)* status: LGTM missing from assignee wernerpe, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @AlexandreAmice)

**Review by wernerpe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376)* status: 1 unresolved discussion, LGTM missing from assignee wernerpe, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @AlexandreAmice)

___
*[`planning/iris/iris_from_clique_cover.cc` line 248 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21376#-NwpFGjH6NK5nFyMmNDm:-NwpFGjH6NK5nFyMmNDn:b-dtg4bz) ([raw file](https://github.com/RobotLocomotion/drake/blob/8e8d4bd0a3dc126a0e5896ce74e157c7c51d0324/planning/iris/iris_from_clique_cover.cc#L248)):*
> ```cpp
>           "Iris builder thread {} failed to compute a set from a clique. This "
>           "clique is being skipped. This can happen if the convex hull of the "
>           "clique contains collisions. The exact error message is {}",
> ```

Maybe we should add a pointer to double checking the collision checker padding:

"Iris builder thread {} failed to compute a set from a clique. This "
 "clique is being skipped. This can happen if the entire clique is closer than configuration_space_margin to an obstacle or convex hull of the "
  "clique contains collisions. The exact error message is {}",



_Code quote:_
```cpp
          "Iris builder thread {} failed to compute a set from a clique. This "
          "clique is being skipped. This can happen if the convex hull of the "
          "clique contains collisions. The exact error message is {}",
```

**Review by wernerpe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376)* status: 1 unresolved discussion, LGTM missing from assignee wernerpe, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @AlexandreAmice)

___
*[`planning/iris/iris_from_clique_cover.cc` line 248 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21376#-NwpFGjH6NK5nFyMmNDm:-NwpSOij5319Dv0hNpzk:b-vp4554) ([raw file](https://github.com/RobotLocomotion/drake/blob/8e8d4bd0a3dc126a0e5896ce74e157c7c51d0324/planning/iris/iris_from_clique_cover.cc#L248)):*
<details><summary><i>Previously, AlexandreAmice (Alexandre Amice) wrote…</i></summary><blockquote>

I chose not to talk  about the collision checker padding because if the obstacle you're close to is a 'configuration_space_obstacle' defined in the `IrisOptions` then padding your `collision_checker` won't fix this.
</blockquote></details>

Wait but there is no stepback for cobstacles, right? https://github.com/RobotLocomotion/drake/blob/master/geometry/optimization/iris.cc#L696
What causes it to throw an error then?

**Comment by RussTedrake** [comment]
> RussTedrake Can you try this on your example that was failing?

It turns out the test that was failing (only on CI and only for a particular flavor of mac), was calling IrisInConfigurationSpace directly. The clique cover test wasn't failing.

**Review by AlexandreAmice — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376)* status: 1 unresolved discussion, LGTM missing from assignee wernerpe, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes

___
*[`planning/iris/iris_from_clique_cover.cc` line 248 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21376#-NwpFGjH6NK5nFyMmNDm:-NwtYdfaBCr7FTjSxpHn:b-5bwfws) ([raw file](https://github.com/RobotLocomotion/drake/blob/8e8d4bd0a3dc126a0e5896ce74e157c7c51d0324/planning/iris/iris_from_clique_cover.cc#L248)):*
<details><summary><i>Previously, wernerpe (Peter Werner) wrote…</i></summary><blockquote>

Wait but there is no stepback for cobstacles, right? https://github.com/RobotLocomotion/drake/blob/master/geometry/optimization/iris.cc#L696
What causes it to throw an error then?
</blockquote></details>

The thrown error happens [here](https://github.com/RobotLocomotion/drake/blob/536165e586153c798a81ddd2d04170486069790a/geometry/optimization/iris.cc#L408). 

Basically, what happens is that you add a hyperplane, take a step back using the configuration space margin, and then your center ends up on the wrong side of the hyperplane.

**Review by wernerpe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376)* status: 1 unresolved discussion, LGTM missing from assignee wernerpe, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @AlexandreAmice)

___
*[`planning/iris/iris_from_clique_cover.cc` line 248 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21376#-NwpFGjH6NK5nFyMmNDm:-Nz5OqqA8u4yXr_v99mM:bd77lzg) ([raw file](https://github.com/RobotLocomotion/drake/blob/8e8d4bd0a3dc126a0e5896ce74e157c7c51d0324/planning/iris/iris_from_clique_cover.cc#L248)):*
<details><summary><i>Previously, AlexandreAmice (Alexandre Amice) wrote…</i></summary><blockquote>

The thrown error happens [here](https://github.com/RobotLocomotion/drake/blob/536165e586153c798a81ddd2d04170486069790a/geometry/optimization/iris.cc#L408). 

Basically, what happens is that you add a hyperplane, take a step back using the configuration space margin, and then your center ends up on the wrong side of the hyperplane.
</blockquote></details>

I am confused, the stepback (called configuration_space_margin) is set to 0 for the cobstacles (configuration_obstacles) if I am not mistaken. This would imply that the above case is not happening?
Here is the line in the code again: https://github.com/RobotLocomotion/drake/blob/master/geometry/optimization/iris.cc#L696

**Review by AlexandreAmice — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376)* status: 1 unresolved discussion, LGTM missing from assignee wernerpe, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @AlexandreAmice)

___
*[`planning/iris/iris_from_clique_cover.cc` line 248 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21376#-NwpFGjH6NK5nFyMmNDm:-NzT_RWo6V0ar4G993qC:bolvczp) ([raw file](https://github.com/RobotLocomotion/drake/blob/8e8d4bd0a3dc126a0e5896ce74e157c7c51d0324/planning/iris/iris_from_clique_cover.cc#L248)):*
<details><summary><i>Previously, wernerpe (Peter Werner) wrote…</i></summary><blockquote>

I am confused, the stepback (called configuration_space_margin) is set to 0 for the cobstacles (configuration_obstacles) if I am not mistaken. This would imply that the above case is not happening?
Here is the line in the code again: https://github.com/RobotLocomotion/drake/blob/master/geometry/optimization/iris.cc#L696
</blockquote></details>

Sure, cspace obstacles won't let  that happen. But task-space obstacles might.

In general, Iris can throw for a lot of reasons, and I don't want to lose 10 regions because one edge case got hit.

**Review by wernerpe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376)* status: 1 unresolved discussion, LGTM missing from assignee wernerpe, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @AlexandreAmice)

___
*[`planning/iris/iris_from_clique_cover.cc` line 248 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21376#-NwpFGjH6NK5nFyMmNDm:-NzTecYz3O6EWrJkK-ht:br7n72q) ([raw file](https://github.com/RobotLocomotion/drake/blob/8e8d4bd0a3dc126a0e5896ce74e157c7c51d0324/planning/iris/iris_from_clique_cover.cc#L248)):*
<details><summary><i>Previously, AlexandreAmice (Alexandre Amice) wrote…</i></summary><blockquote>

Sure, cspace obstacles won't let  that happen. But task-space obstacles might.

In general, Iris can throw for a lot of reasons, and I don't want to lose 10 regions because one edge case got hit.
</blockquote></details>

Yes, agreed. But I stand by my point it is not the manually added c-obstacles that are causing this crash. I dont see how they could - the step back is not required as they are assumed to be convex and is set to zero. We should get the try catch merged though.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376)* status: 1 unresolved discussion, LGTM missing from assignee wernerpe, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @AlexandreAmice)

___
*[`planning/iris/iris_from_clique_cover.cc` line 241 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21376#-NzTb5WE6D2o3htlHVCt:-NzTb5WFD0QGqC5RFs_G:b-5yebkf) ([raw file](https://github.com/RobotLocomotion/drake/blob/8e8d4bd0a3dc126a0e5896ce74e157c7c51d0324/planning/iris/iris_from_clique_cover.cc#L241)):*
> ```cpp
>       iris_region.emplace(IrisInConfigurationSpace(
>           checker.plant(), checker.plant_context(builder_id), iris_options));
>     } catch (const std::logic_error& e) {
> ```

BTW Catching exceptions is forbidden:

https://drake.mit.edu/styleguide/cppguide.html#Exceptions

Possibly platform review will allow it as a temporary style violation as a step along the way to a better solution that doesn't catch exceptions, but for that approach at minimum you'd need to lay out the plan for how we'll eventually get rid of the `catch`.

**Review by AlexandreAmice — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376)* status: 1 unresolved discussion, LGTM missing from assignee wernerpe, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes

___
*[`planning/iris/iris_from_clique_cover.cc` line 241 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21376#-NzTb5WE6D2o3htlHVCt:-NzTnvF30kjQbl746oAn:b-u3tw4m) ([raw file](https://github.com/RobotLocomotion/drake/blob/8e8d4bd0a3dc126a0e5896ce74e157c7c51d0324/planning/iris/iris_from_clique_cover.cc#L241)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Catching exceptions is forbidden:

https://drake.mit.edu/styleguide/cppguide.html#Exceptions

Possibly platform review will allow it as a temporary style violation as a step along the way to a better solution that doesn't catch exceptions, but for that approach at minimum you'd need to lay out the plan for how we'll eventually get rid of the `catch`.
</blockquote></details>

What would be your recommended way of handling this case then?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21376)* status: 1 unresolved discussion, LGTM missing from assignee wernerpe, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @AlexandreAmice)

___
*[`planning/iris/iris_from_clique_cover.cc` line 241 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21376#-NzTb5WE6D2o3htlHVCt:-NzToRlMEMaxEuobHRIO:budj0te) ([raw file](https://github.com/RobotLocomotion/drake/blob/8e8d4bd0a3dc126a0e5896ce74e157c7c51d0324/planning/iris/iris_from_clique_cover.cc#L241)):*
<details><summary><i>Previously, AlexandreAmice (Alexandre Amice) wrote…</i></summary><blockquote>

What would be your recommended way of handling this case then?
</blockquote></details>

If we want to programmatically handle errors from `IrisInConfigurationSpace` then it would need to _return_ the error information instead of throwing.

**Comment by jwnimmer-tri** [comment]
Given recent developments like the new objectives in #21821, is this PR still relevant?  If not, then please close it to conserve CI resources.
